app/human-detector-app/auth/KeyCloakAuth.ts

```diff
@@ -0,0 +1,92 @@
+import { makeRedirectUri, useAuthRequest } from 'expo-auth-session';
+import * as AuthSession from 'expo-auth-session';
+import * as CryptoJS from 'crypto-js';
+
+function genCodeVerifier(size: number) {
+  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
+  const charLength = chars.length;
+  let result = '';
+  for (let i = 0; i < size; i++) {
+    result += chars.charAt(Math.floor(Math.random() * charLength));
```

Comment:
Apparently `Math.random()` [doesn't provide cryptographically secure random numbers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random#sect1). The MDN docs recommend using `window.crypto.getRandomValues()`, but you probably can't access that from Expo so we could outsource the string generation process to something like [crypto-random-string](https://www.npmjs.com/package/crypto-random-string)

Reply:
This seems to be having the same problem that I was having when trying to use other packages for random strings.  It uses a node package for its generation which expo won't be able to use (thanks expo).  I can't find an expo equivalent for it in the docs which has gotten me a little stuck.

Reply:
Gotta love Expo. Looks like they have their own [Random API](https://docs.expo.dev/versions/latest/sdk/random/#randomgetrandombytesbytecount), but that could produce invalid characters, so a hacky but probably functional way might be to base64-encode the generated bytes? You would need to replace any `+` and `/` characters in the resulting string with one of the allowed punctuation characters ([ref](https://www.oauth.com/oauth2-servers/pkce/authorization-request/)) though. I'm fine leaving it as-is for now, seems like it's a pain to do this stuff with Expo